### PR TITLE
add replaceWithSeveral

### PR DESCRIPTION
**Pull request by matozoid**
I'm not sure the argument should be a list. Maybe a vararg. Not sure how this will be used in the future.

### CODE REVIEW

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt [hunk 1]
```diff
@@ -61,14 +61,14 @@ private fun providesNodes(kclass: KClass<*>?): Boolean {
 }
 
 /**
- * @return can this class be considered an AST node?
+ * @return can [this] class be considered an AST node?
```
Comment (ftomassetti) [inline src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: why the square brackets?

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt [hunk 2]
```diff
@@ -29,7 +29,7 @@ fun Node.assignParents() {
 }
 
 /**
- * Recursively execute "operation" on this node, and all nodes below this node.
+ * Recursively execute [operation] on [this] node, and all nodes below this node.
```
Comment (ftomassetti) [inline src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: why the square brackets?

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt [hunk 3]
```diff
@@ -272,12 +272,45 @@ fun Node.mapTree(operation: (Node) -> Node): Node {
...
         throw IllegalStateException("Parent not set")
     }
     this.parent!!.transformTree { if (it == this) other else it }
 }
+
+/**
+ * Looks for [oldNode] in the lists of nodes in this node.
+ * When found, it is removed, and in its place the [newNodes] are inserted.
+ */
+@Suppress("UNCHECKED_CAST") // assumption: a MutableList with a Node in it is a MutableList<Node>
```
Comment (ftomassetti) [inline src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: What happens if the property allows for one element and not a list?

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt [hunk 4]
```diff
@@ -272,12 +272,45 @@ fun Node.mapTree(operation: (Node) -> Node): Node {
...
+fun Node.replaceWithSeveral(oldNode: Node, newNodes: List<Node>) {
+    relevantMemberProperties().forEach { property ->
+        when (val value = property.get(this)) {
+            is MutableList<*> -> {
```
Comment (ftomassetti) [inline src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: what if the value is the single node we were looking for? Should we throw an exception?

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt (hunk 1 above)
Comment (matozoid) [reply src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: It's a link, check out https://kotlinlang.org/docs/reference/kotlin-doc.html

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt (hunk 3 above)
Comment (matozoid) [reply src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: Nothing. Is that a valid use case or shall we make it an error?

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt (hunk 4 above)
Comment (matozoid) [reply src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: Oh, we're asking each other the same thing :-)

Comment (matozoid) [reply src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: For my use case it's an error, so I'll make it an error.

File: src/main/kotlin/com/strumenta/kolasu/model/Processing.kt (hunk 3 above)
Comment (ftomassetti) [reply src/main/kotlin/com/strumenta/kolasu/model/Processing.kt]: I would expect it to throw an error, it would be more clear to me

**Comment by matozoid** [comment]
Is this okay, @ftomassetti ?

**Comment by ftomassetti** [comment]
Looks great, thanks!

**Pull request merged**